lib/ibm/cloud/sdk/vpc/instances.rb

```diff
@@ -38,21 +38,115 @@ def params(start: nil, limit: nil, name: nil, vpc_id: nil, vpc_crn: nil, vpc_nam
+          def status
+            @data[:status]
+          end
+
+          # Whether the state of the VM is in failed state.
+          # @return [Boolean]
+          def failed?
+            status == ERROR_STATE
+          end
+
+          # Whether the state of the VM is in the started state.
+          # @return [Boolean]
+          def started?
+            status == RUNNING_STATE
+          end
+
+          # Whether the state of the VM is in a stopped or paused state.
+          # @return [Boolean]
+          def stopped?
+            STOPPED_STATES.include?(status)
+          end
+
+          # Whether the state of the VM is in a transitional state.
+          # @return [Boolean]
+          def transitional?
+            TRANSITIONAL_STATES.include?(status)
+          end
+
+          # Interact with instance actions.
+          # @return [INSTANCE::Actions]
           def actions
             INSTANCE::Actions.new(self)
           end
 
+          # Interact with instance network interfaces.
+          # @return [INSTANCE::NetworkInterfaces]
           def network_interfaces
             INSTANCE::NetworkInterfaces.new(self)
           end
 
+          # Interact with instance volume attachements.
+          # @return [INSTANCE::VolumeAttachments]
           def volume_attachments
             INSTANCE::VolumeAttachments.new(self)
           end
 
+          # Return the data used for initializing this VM.
           def initialization
             adhoc(method: 'get', path: 'initialization').json
           end
+
+          # Wait for the VM instance to be in a stable state.
+          # @param started [Boolean] When true wait for started, when false wait for stopped.
+          # @param sleep_time [Integer] The time to sleep between refreshes.
+          # @param timeout [Integer] The number of seconds before raising an error.
+          # @param block [Proc] A block to test against. Must return a boolean.
+          # @return [Boolean] The return of the status check.
+          # @raise [RuntimeError] Instance goes into failed state.
+          # @raise [RuntimeError] Timeout has been reached.
+          def wait_for(started: true, sleep_time: 5, timeout: 600, &block)
+            @logger.info("Starting wait for, instance #{id} starts in state #{status}.")
+            until test_state(started: started, &block)
+              raise "VM #{id} is in a failed state." if failed?
+
+              timeout = sleep_counter(sleep_time, timeout)
+              raise "Time out while waiting #{id} to be stable." if timeout <= 0
+            end
+            @logger.info("Finished wait for instance #{id} ends in state #{status}.")
+            test_state(&block)
+          end
+
+          private
+
+          # Used with wait_for return true if started is true and vm is started, or started is false and vm is stopped.
+          # @return [Boolean]
+          def test_state(started: true, &block)
```

Comment:
Hm combining test_state and a block seems to be extra complexity when you can accomplish everything with
```
def wait_for(sleep_time: 5, timeout: 600, &block)
  loop do
    refresh
    raise "VM #{id} is in a failed state." if failed?
    break if block.call(self)
    timeout = sleep_counter(sleep_time, timeout)
    raise "Time out while waiting #{id} to be stable." if timeout <= 0
  end
end
```

And the difference to the caller in the simple case is `wait_for(started: true)` vs `wait_for(&:started?)`

I won't hold this up for this but I think we could simplify this a lot without losing anything

Reply:
I was debating on removing started and had pretty much the same syntax. I don't like having options that can be silently overridden by other options. I deferred to the more complicated since using started allows people less familiar with blocks to use wait_for for the simplest reasons can just pass in an option. I guess another option is to add two more methods wait_for_started and wait_for_stopped each passing in the requisite block in its body. That is probably a better solution than the started option. I'll update.